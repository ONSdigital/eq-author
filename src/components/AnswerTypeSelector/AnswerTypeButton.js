import React from "react";
import PropTypes from "prop-types";
import { IconGridButton } from "components/IconGrid";
import { keys } from "lodash";

import checkboxIcon from "./icons/checkbox.svg";
import currencyIcon from "./icons/currency.svg";
import dateIcon from "./icons/date.svg";
import numberIcon from "./icons/number.svg";
import radioIcon from "./icons/radio.svg";
import selectIcon from "./icons/select.svg";
import textareaIcon from "./icons/textarea.svg";
import textfieldIcon from "./icons/textfield.svg";
import timeIcon from "./icons/time.svg";

export const icons = {
  checkbox: checkboxIcon,
  currency: currencyIcon,
  date: dateIcon,
  number: numberIcon,
  radio: radioIcon,
  select: selectIcon,
  textarea: textareaIcon,
  textfield: textfieldIcon,
  time: timeIcon
};
export default class AnswerTypeButton extends React.Component {
  static propTypes = {
    type: PropTypes.oneOf(keys(icons)).isRequired,
    title: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
    disabled: PropTypes.bool
  };

  handleClick = () => {
    this.props.onClick(this.props.type);
  };

  render() {
    return (
      <IconGridButton
        disabled={this.props.disabled}
        iconSrc={icons[this.props.type]}
        onClick={this.handleClick}
        title={this.props.title}
      />
    );
  }
}
