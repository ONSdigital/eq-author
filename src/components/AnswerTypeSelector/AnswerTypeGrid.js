/* eslint-disable react/no-find-dom-node */
import React from "react";
import { findDOMNode } from "react-dom";
import PropTypes from "prop-types";
import { IconGrid } from "components/IconGrid";
import AnswerTypeButton from "./AnswerTypeButton";
import styled from "styled-components";

const Menu = styled.div`
  background-color: white;
  box-shadow: rgba(0, 0, 0, 0.16) 0 5px 20px 0px;
  width: 340px;
`;

const Header = styled.div`
  padding-left: 1em;
  padding-right: 1em;
  padding-top: 1.5em;
`;

const Title = styled.div`
  color: #444;
  font-size: 0.6em;
  font-weight: bold;
  text-transform: uppercase;
`;

export const CloseButton = styled.button`
  color: #444;
  border: none;
  background: none;
  font-size: 1em;
  position: absolute;
  top: 1.5rem;
  right: 1rem;
  line-height: 0.6;
  padding: 0;
  cursor: pointer;
`;

class AnswerTypeGrid extends React.Component {
  static propTypes = {
    onClose: PropTypes.func,
    onSelect: PropTypes.func.isRequired,
    "aria-labelledby": PropTypes.string
  };

  handleSelect = type => {
    this.props.onClose();
    this.props.onSelect(type);
  };

  focusMenuItem = () => {
    findDOMNode(this.button).focus();
  };

  saveButtonRef = button => {
    this.button = button;
  };

  render() {
    const { onClose, "aria-labelledby": labelledby } = this.props;

    return (
      <Menu>
        <Header>
          <Title>Answer Type</Title>
          <CloseButton onClick={onClose} aria-label="Close" type="button">
            Ã—
          </CloseButton>
        </Header>
        <IconGrid aria-labelledby={labelledby}>
          <AnswerTypeButton
            type="Checkbox"
            title="Checkbox"
            onClick={this.handleSelect}
            ref={this.saveButtonRef}
          />
          <AnswerTypeButton
            type="Radio"
            title="Radio"
            onClick={this.handleSelect}
          />
          <AnswerTypeButton
            type="TextField"
            title="Text"
            onClick={this.handleSelect}
          />
          <AnswerTypeButton
            type="TextArea"
            title="Textarea"
            onClick={this.handleSelect}
            disabled
          />
          <AnswerTypeButton
            type="Currency"
            title="Currency"
            onClick={this.handleSelect}
          />
          <AnswerTypeButton
            type="Number"
            title="Number"
            onClick={this.handleSelect}
          />
          <AnswerTypeButton
            type="Date"
            title="Date"
            onClick={this.handleSelect}
            disabled
          />
          <AnswerTypeButton
            type="Time"
            title="Time"
            onClick={this.handleSelect}
            disabled
          />
          <AnswerTypeButton
            type="Select"
            title="Select list"
            onClick={this.handleSelect}
            disabled
          />
        </IconGrid>
      </Menu>
    );
  }
}

export default AnswerTypeGrid;
