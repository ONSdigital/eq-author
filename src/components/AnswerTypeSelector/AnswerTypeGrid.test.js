import React from "react";
import { shallow } from "enzyme";
import AnswerTypeGrid, { CloseButton } from "./AnswerTypeGrid";
import AnswerTypeButton from "./AnswerTypeButton";

let wrapper, handleClose, handleSelect;

describe("components/AnswerTypeGrid", () => {
  beforeEach(() => {
    handleClose = jest.fn();
    handleSelect = jest.fn();

    wrapper = shallow(
      <AnswerTypeGrid
        onClose={handleClose}
        aria-labelledby="foo"
        onSelect={handleSelect}
      />
    );
  });

  it("should render", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("when close button clicked", () => {
    it("should invoke `onClose` callback", () => {
      wrapper.find(CloseButton).simulate("click");
      expect(handleClose).toHaveBeenCalled();
    });
  });

  describe("when answer type button clicked", () => {
    it("should invoke `onSelect` callback", () => {
      const answerType = "TextField";
      wrapper
        .find(AnswerTypeButton)
        .first()
        .simulate("click", answerType);

      expect(handleClose).toHaveBeenCalled();
      expect(handleSelect).toHaveBeenCalledWith(answerType);
    });
  });

  it("should focus on the underlying button element when focusMenuItem is called", () => {
    const btn = document.createElement("button");
    wrapper.instance().saveButtonRef(btn);
    wrapper.instance().focusMenuItem();
    expect(document.activeElement).toBe(btn);
  });
});
