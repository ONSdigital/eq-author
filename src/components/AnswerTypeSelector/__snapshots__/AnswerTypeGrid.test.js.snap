// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AnswerTypeGrid should render 1`] = `
<AnswerTypeGrid__Menu>
  <AnswerTypeGrid__Header>
    <AnswerTypeGrid__Title>
      Answer Type
    </AnswerTypeGrid__Title>
    <AnswerTypeGrid__CloseButton
      aria-label="Close"
      onClick={[Function]}
      type="button"
    >
      Ã—
    </AnswerTypeGrid__CloseButton>
  </AnswerTypeGrid__Header>
  <IconGrid
    aria-labelledby="foo"
  >
    <AnswerTypeButton
      onClick={[Function]}
      title="Checkbox"
      type="Checkbox"
    />
    <AnswerTypeButton
      disabled={true}
      onClick={[Function]}
      title="Radio"
      type="Radio"
    />
    <AnswerTypeButton
      onClick={[Function]}
      title="Text"
      type="TextField"
    />
    <AnswerTypeButton
      disabled={true}
      onClick={[Function]}
      title="Textarea"
      type="TextArea"
    />
    <AnswerTypeButton
      disabled={true}
      onClick={[Function]}
      title="Currency"
      type="Currency"
    />
    <AnswerTypeButton
      disabled={true}
      onClick={[Function]}
      title="Number"
      type="Number"
    />
    <AnswerTypeButton
      disabled={true}
      onClick={[Function]}
      title="Date"
      type="Date"
    />
    <AnswerTypeButton
      disabled={true}
      onClick={[Function]}
      title="Time"
      type="Time"
    />
    <AnswerTypeButton
      disabled={true}
      onClick={[Function]}
      title="Select list"
      type="Select"
    />
  </IconGrid>
</AnswerTypeGrid__Menu>
`;
